### DISCUSSION TITLE
MCU becoming unresponsive when setting RTC flags

### DISCUSSION THREAD

**Discussion by InstantMuffin** [SpenceKonde/megaTinyCore#747] [Q&A]
Hello,

Trying to find out some tiny details about how the MCU behaves and also simplifying going into powerdown mode (for me), I have come across an oddity that someone with more knowledge might clear up.
At first I suspected this to be an issue with a specific board/setup I had, however this just occurred using the minimal config as well (1627, 100+10nf caps, reset circuitry with autoreset as described in the documentation, using optiboot, plus resistor and the led).

The following setup with this code gives me the following output:

```
#include <avr/sleep.h>

/* Period select */
enum PowerDownDuration
{
  PowerDownDuration_125us = (0x01<<3),  /* RTC Clock Cycles 4 */
  PowerDownDuration_250us = (0x02<<3),  /* RTC Clock Cycles 8 */
  PowerDownDuration_500us = (0x03<<3),  /* RTC Clock Cycles 16 */
  PowerDownDuration_1ms = (0x04<<3),  /* RTC Clock Cycles 32 */
  PowerDownDuration_2ms = (0x05<<3),  /* RTC Clock Cycles 64 */
  PowerDownDuration_4ms = (0x06<<3),  /* RTC Clock Cycles 128 */
  PowerDownDuration_8ms = (0x07<<3),  /* RTC Clock Cycles 256 */
  PowerDownDuration_16ms = (0x08<<3),  /* RTC Clock Cycles 512 */
  PowerDownDuration_31ms = (0x09<<3),  /* RTC Clock Cycles 1024 */
  PowerDownDuration_62ms = (0x0A<<3),  /* RTC Clock Cycles 2048 */
  PowerDownDuration_125ms = (0x0B<<3),  /* RTC Clock Cycles 4096 */
  PowerDownDuration_250ms = (0x0C<<3),  /* RTC Clock Cycles 8192 */
  PowerDownDuration_500ms = (0x0D<<3),  /* RTC Clock Cycles 16384 */
  PowerDownDuration_1000ms = (0x0E<<3),  /* RTC Clock Cycles 32768 */
};

void PreconfigurePowerDown(PowerDownDuration cycles)
{
  while (RTC.STATUS > 0)
  {
    ;
  }
  RTC.CLKSEL = RTC_CLKSEL_INT32K_gc;

  RTC.PITINTCTRL = RTC_PI_bm;

  RTC.PITCTRLA = ((RTC_PERIOD_enum)(cycles)) | RTC_PITEN_bm;
}

ISR(RTC_PIT_vect)
{
  RTC.PITINTFLAGS = RTC_PI_bm;
}

void EnterPowerDownMode(bool turnOffADC = true)
{
  if(turnOffADC)
  {
    ADC0.CTRLA &= ~ADC_ENABLE_bm;
  }
  
  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();
  sleep_cpu();

  if(turnOffADC)
  {
    ADC0.CTRLA |= ADC_ENABLE_bm;
  }
}

void EnterPowerDownMode(PowerDownDuration cycles, bool turnOffADC = true)
{
  PreconfigurePowerDown(cycles);

  EnterPowerDownMode(turnOffADC);
}

void setup()
{
  PreconfigurePowerDown(PowerDownDuration_1000ms);
  pinMode(8, OUTPUT);  
  
  Serial.pins(18, 19);
  Serial.begin(9600);
}

void loop()
{
  Serial.println("HIGH");
  Serial.flush();

  digitalWrite(8, HIGH);
  unsigned long startTime = micros();
  //Comment this line out and it will work
  PreconfigurePowerDown(PowerDownDuration_1000ms);
  unsigned long finishTime = micros();

  Serial.println("Config time in microseconds:");
  Serial.println(finishTime - startTime);
  Serial.flush();
  delay(11);

  Serial.println("Sleep");
  Serial.flush();

  EnterPowerDownMode(true);

  Serial.println("Wakeup");
  Serial.flush();

  Serial.println("LOW");
  Serial.flush();

  digitalWrite(8, LOW);
  delay(1000);
}
```

```HIGH
Config time in microseconds:
6
HIGH
Config time in microseconds:
6
Sleep
Wakeup
LOW
HIGH
```

After that the MCU just becomes unresponsive.
I added an LED to make sure that it's not just the serial connection breaking, however, the LED stays on indefinitely. It does seem the MCU is frozen.

I noticed when commenting out this line in loop()
```
//Comment this line out and it will work
PreconfigurePowerDown(PowerDownDuration_1000ms);
```

it will work as expected without becoming unresponsive.

I also tried using this script without turning the adc on and off, which made no difference.
This seems rather odd to me, and I actually have no idea what is going on here. I didn't see any mentioning of nitpicks in the PowerSave documentation in this repo and would love to have shed some light on this.

And forgive me, but I'm tagging you @SpenceKonde since you're probably the only one who knows what is happening here anyway.

**Comment by InstantMuffin** [comment]
After the first sleep cycle (PreconfigurePowerDown and EnterPowerDownMode call), CNTBUSY flag is constantly set.
Changing the while condition to

```
while (RTC.STATUS != 2 && RTC.STATUS != 0)
{
  ;
}
```

allows execution and so far I have not had any issues running the altered code for a couple of minutes.

Since the code does not write to the RTC.COUNT register this should be fine.

Maybe someone has a technical explanation for this. Maybe I'm doing something I'm not supposed to do here, or maybe the documentation should address this.

**Comment by SpenceKonde** [comment]
What chip is this? Are you trying to run either the RTC or PIT without the other one on a tinyAVR 0 or 1-series part that's not a 3216 or 3217? 

See the errata - you have to either run both of them or neither of them, or they don't work correctly and the status shows that they're eternally syncing

This issue should be fixed on the 2-series and Dx-series though. and if they ever give us a damned silicon rev. I would gladly pay an extra 10-20% for their products if they;d actually fix the bugs and give us new silicon revs.

**Reply by InstantMuffin** [reply]
> What chip is this?

1627, as I mentioned earlier.

> Are you trying to run either the RTC or PIT without the other one on a tinyAVR 0 or 1-series part that's not a 3216 or 3217?

I'm doing nothing other than running this sketch.
The sketch is based on the code in the PowerSave.md example, the differences are the wrapper functions for what used to be "RTC_init()" and renaming the RTC_PERIOD_enum into something that represents the powerdown duration (given I'm just using the 32KHz RTC).

You can reproduce this by using the PowerSave.md example and calling "RTC_init()" twice.
Essentially if you run the code example of PowerSave.md the STATUS register will always be "2" after the first "RTC_init()" call.
The idea is (at least for me / what I'm trying to do) to have a wrapper that presets different powerdown durations for me to set in a readable format (me knowing the time immediately) and switching in between them.

So far I'm just using the altered STATUS condition check and kept working on it and now have a wrapper that does the above plus switching between the 32KHz and 1KHz oscillator for more available powerdown durations plus keeping millis() accurate by advancing it properly and accounting for leap milliseconds happening when I accumulate multiple microsecond-powerdowns.

The very rough current-state that I'm working with:
```
#include <avr/sleep.h>

struct PowerDownSettings
{
  RTC_CLKSEL_t Prescaler;
  RTC_PERIOD_t ClockPeriod;
  unsigned int SleepMillis;
  unsigned int SleepMicros;
};

//21
static const PowerDownSettings POWERDOWNSETTINGS[] = {
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC4_gc, 0, 125},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC8_gc, 0, 250},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC16_gc, 0, 500},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC32_gc, 1, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC64_gc, 2, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC128_gc, 4, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC256_gc, 8, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC512_gc, 16, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC1024_gc, 31, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC2048_gc, 62, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC4096_gc, 125, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC8192_gc, 250, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC16384_gc, 500, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_CYC32768_gc, 1000, 0},
  {RTC_CLKSEL_INT1K_gc, RTC_PERIOD_CYC1024_gc, 1024, 0},
  {RTC_CLKSEL_INT1K_gc, RTC_PERIOD_CYC2048_gc, 2048, 0},
  {RTC_CLKSEL_INT1K_gc, RTC_PERIOD_CYC4096_gc, 4096, 0},
  {RTC_CLKSEL_INT1K_gc, RTC_PERIOD_CYC8192_gc, 8192, 0},
  {RTC_CLKSEL_INT1K_gc, RTC_PERIOD_CYC16384_gc, 16384, 0},
  {RTC_CLKSEL_INT1K_gc, RTC_PERIOD_CYC32768_gc, 32768, 0},
  {RTC_CLKSEL_INT32K_gc, RTC_PERIOD_OFF_gc, 0, 0},
};

enum PowerDownDuration
{
  //32KHz Oscillator
  PowerDownDuration_125us = 0,  /* RTC Clock Cycles 4 */
  PowerDownDuration_250us,      /* RTC Clock Cycles 8 */
  PowerDownDuration_500us,      /* RTC Clock Cycles 16 */
  PowerDownDuration_1ms,        /* RTC Clock Cycles 32 */
  PowerDownDuration_2ms,        /* RTC Clock Cycles 64 */
  PowerDownDuration_4ms,        /* RTC Clock Cycles 128 */
  PowerDownDuration_8ms,        /* RTC Clock Cycles 256 */
  PowerDownDuration_16ms,       /* RTC Clock Cycles 512 */
  PowerDownDuration_31ms,       /* RTC Clock Cycles 1024 */
  PowerDownDuration_62ms,       /* RTC Clock Cycles 2048 */
  PowerDownDuration_125ms,      /* RTC Clock Cycles 4096 */
  PowerDownDuration_250ms,      /* RTC Clock Cycles 8192 */
  PowerDownDuration_500ms,      /* RTC Clock Cycles 16384 */
  PowerDownDuration_1000ms,     /* RTC Clock Cycles 32768 */
  //1KHz Oscillator
  PowerDownDuration_1024ms,     /* RTC Clock Cycles 1024 */
  PowerDownDuration_2048ms,     /* RTC Clock Cycles 2048 */
  PowerDownDuration_4096ms,     /* RTC Clock Cycles 4096 */
  PowerDownDuration_8192ms,     /* RTC Clock Cycles 8192 */
  PowerDownDuration_16384ms,    /* RTC Clock Cycles 16384 */
  PowerDownDuration_32768ms,    /* RTC Clock Cycles 32768 */

  //32KHz Oscillator
  //millis() won't work here
  PowerDownDuration_Forever,    /* RTC_PERIOD_OFF_gc */
};

PowerDownDuration powerDownSettingsID = PowerDownDuration_125us;

unsigned int sleptMicros = 0;
unsigned int leapMillis = 0;

void PreconfigurePowerDown(PowerDownDuration duration)
{
  while (RTC.STATUS != 0 && RTC.STATUS != 2)
  {
    ;
  }

  RTC.CLKSEL = POWERDOWNSETTINGS[duration].Prescaler;

  RTC.PITINTCTRL = RTC_PI_bm;

  RTC.PITCTRLA = (POWERDOWNSETTINGS[duration].ClockPeriod) | RTC_PITEN_bm;
}

ISR(RTC_PIT_vect)
{
  RTC.PITINTFLAGS = RTC_PI_bm;
}

void EnterPowerDownMode(bool turnOffADC = true)
{
  if(turnOffADC)
  {
    ADC0.CTRLA &= ~ADC_ENABLE_bm;
  }
  
  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  sleep_enable();
  sleep_cpu();

  sleptMicros += POWERDOWNSETTINGS[powerDownSettingsID].SleepMicros;
  if(sleptMicros >= 1000)
  {
    leapMillis = (sleptMicros / 1000);
    sleptMicros -= leapMillis * 1000;
  }
  else
  {
    leapMillis = 0;
  }

  stop_millis();
  set_millis(millis() + POWERDOWNSETTINGS[powerDownSettingsID].SleepMillis + leapMillis);
  restart_millis();

  if(turnOffADC)
  {
    ADC0.CTRLA |= ADC_ENABLE_bm;
  }
}

void EnterPowerDownMode(PowerDownDuration duration, bool turnOffADC = true)
{
  PreconfigurePowerDown(duration);

  EnterPowerDownMode(turnOffADC);
}

void setup()
{
  Serial.pins(18, 19);
  Serial.begin(9600);
  
  pinMode(8, OUTPUT);  
}

void loop()
{
  Serial.println("HIGH");
  Serial.flush();

  digitalWrite(8, HIGH);

  PreconfigurePowerDown(PowerDownDuration_1000ms);

  Serial.println("Sleep");
  Serial.flush();

  EnterPowerDownMode(true);

  Serial.println("Wakeup");
  Serial.flush();

  Serial.println("LOW");
  Serial.flush();

  digitalWrite(8, LOW);
  delay(1000);
}
```

**Reply by InstantMuffin** [reply]
Ah, I see what you mean.
https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Errata.md#disabling-the-rtc-stops-the-pit

How should one handle this? Because the PowerSave.md code example is going to run into this without addressing or even mentioning this (the issue is being mentioned but not that the provided code sample runs into this). And I was introspective enough to assume that **I** am doing something wrong and not the multi-billion manufacturing company and therefore not checking the errata sheet.

**Comment by SpenceKonde** [comment]
I think you just have to not turn off the RTC... Also, I'll note that each of the status bits refers specifically to one register - other registers should still be writable, most importantly ctrla of whichever one is off to turn it on so the other one will work. And remember that every time you write to the RTC.CTRLA on bug impacted parts (all 1-series and 0 series except 3216 and 3217 die rev C, it resets the prescaler, extending the current RTC "tick" 

But I wouldn't expect that on a 1624 - that's a 2-series, whereas the RTC bug is one of the few that was actually fixed with a die rev on some parts, and the fixed parts were available before the 2-series product brief even came out. Usually they seem to just add the problems to the errata and never revise to die to fix them - butcha can't assume they won't fix them either. 


Considering how bad the errata on the DA and tinyAVR 0/1-series is, I think they owe it to us to have an AVR errata chart listing every silicon bug known, which parts it is known to be on, which parts it is known not to be on, what the priority is (ie, "unlikely to be fixced in this product but is corrected in new releases since mm/dd/yy " "low priority, will correct if someone is already working in a section of the design and the fix is easy" "top priority" and the status, ex "planned fix has not been built into any silicon yet" "planned fix has been in parts released since mm/dd/yy but has not been ported to this device" "Fix is ready to go and will be in next die revision, planned for first half of 2023." up to "parts with fix went into mass production from mm/dd/yy". Or "We recognize that this behavior is really awful, but it's going to take a lot of engineering effort to correct, and customers should plan to work around this issue for the forseeable future". or "nobody can agree on what the correct behavior here is, This

**Reply by InstantMuffin** [reply]
> But I wouldn't expect that on a 1624

Odd isn't it?
I bought these 1627s off of a reputable German distributor - reichelt.
I don't know if you have any yourself and they don't have this bug, maybe it just affects a specific batch. If you like I can photograph the MCU under a microscope so you have a batch number.
That said if I was from the US as well I'd mail you one for keeps.

(Oh I think plain mail would do and costs like a buck for me, supposedly with only 7 days of shipping time.)

**Reply by InstantMuffin** [reply]
> I think you just have to not turn off the RTC

I apologize for the follow-up. But where in the code am I turning off the RTC?
Thank you for all the help so far and for all the work you have done with those cores.

**Comment by SpenceKonde** [comment]
I have not tested mine for the bug. don't really use the RTC much and there are more pressing concerns.
